test-utils/transaction-factory/src/lib.rs

```diff
@@ -0,0 +1,192 @@
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use log::info;
+use rand::{Rng, SeedableRng};
+use rand::rngs::StdRng;
+
+use balances;
+use balances::Call as BalancesCall;
+use client::block_builder::api::BlockBuilder;
+use client::runtime_api::ConstructRuntimeApi;
+use consensus_common::{BlockOrigin, ImportBlock, ForkChoiceStrategy};
+use consensus_common::block_import::BlockImport;
+use indices;
+use keyring::sr25519::Keyring;
+use node_primitives::Hash;
+use node_runtime::{Call, CheckedExtrinsic, UncheckedExtrinsic};
+use parity_codec::{Decode, Encode};
+use primitives::sr25519;
+use primitives::crypto::Pair;
+use sr_primitives::generic::Era;
+use sr_primitives::traits::{As, Block, Header, ProvideRuntimeApi};
+use substrate_service::{FactoryBlock, FactoryFullConfiguration, FullClient, new_client, ServiceFactory};
+
+/// Generates a random `AccountId` from `seed`.
+fn gen_random_account_id(seed: u64) -> node_primitives::AccountId {
+	let mut rng: StdRng = SeedableRng::seed_from_u64(seed);
+
+	let mut seed_bytes = [0u8; 32];
+	for i in 0..32 {
+		seed_bytes[i] = rng.gen::<u8>();
+	}
+
+	let pair: sr25519::Pair = sr25519::Pair::from_seed(seed_bytes);
+	pair.public().into()
+}
+
+/// Creates an `UncheckedExtrinsic` containing the appropriate signature for
+/// a `CheckedExtrinsics`.
+fn sign<F: ServiceFactory>(xt: CheckedExtrinsic, prior_block_hash: Hash, block_no: u64) -> UncheckedExtrinsic {
+	let mut phase = 0;
+	if block_no > 0 {
+		phase = block_no - 1;
+	}
+
+	match xt.signed {
+		Some((signed, index)) => {
+			let era = Era::mortal(256, phase);
+			let payload = (index.into(), xt.function, era, prior_block_hash);
+			let key = Keyring::from_public(&signed)
+				.expect("Failed to create public key from signed");
+			let signature = payload.using_encoded(|b| {
+				if b.len() > 256 {
+					key.sign(&sr_io::blake2_256(b))
+				} else {
+					key.sign(b)
+				}
+			}).into();
+			UncheckedExtrinsic {
+				signature: Some((indices::address::Address::Id(signed), signature, payload.0, era)),
+				function: payload.1,
+			}
+		}
+		None => UncheckedExtrinsic {
+			signature: None,
+			function: xt.function,
+		},
+	}
+}
+
+/// Manufacture `num` transactions from Alice to `num` accounts.
+pub fn factory<F>(
+	config: FactoryFullConfiguration<F>,
+	num: u64,
+) -> cli::error::Result<()>
+where
+	F: ServiceFactory,
+	F::RuntimeApi: ConstructRuntimeApi<FactoryBlock<F>, FullClient<F>>,
+	FullClient<F>: ProvideRuntimeApi,
+	<FullClient<F> as ProvideRuntimeApi>::Api: BlockBuilder<FactoryBlock<F>>,
+{
+	info!("Creating {} transactions...", num);
+
+	let client = new_client::<F>(&config)?;
+
+	let mut prior_block_hash = client.best_block_header()?.hash();
+
+	// TODO extract timestamp from last block inherents via the api
+	// 	let api = client.runtime_api(); ?
+	// 	let block = client.block(&BlockId::Hash(prior_block_hash))?;
+	let now = SystemTime::now();
+	let mut last_ts: u64 = now.duration_since(UNIX_EPOCH)
+		.expect("Time went backwards").as_secs();
```

Comment:
I think the format of `expect` proofs are more accurately `expect("proof; qed")`.

```suggestion
		.expect("Time cannot go backward; qed").as_secs();
```

Reply:
Or maybe not in the client code.

Reply:
Is this proof plausible tho?

Time (at least in computer science) can and do go backwards. `Instant` for example provides a property that the time either goes forward or doesn't go : )
However, it looks like we use `SystemTime` here, which is says the following in its documentation:

> Distinct from the `Instant` type, this time measurement *is not monotonic*. This means that you can save a file to the file system, then save another file to the file system, *and the second file has a SystemTime measurement earlier than the first*. In other words, an operation that happens after another operation in real time may have an earlier SystemTime!

So this proof or the suggested proof doesn't look plausible taking in mind this note.

Reply:
But in this case we use the constant `UNIX_EPOCH` to compare against `now`.

As I understand it `now.duration_since(UNIX_EPOCH)` would only fail if `now` is smaller than `1970-01-01 00:00:00 UTC`, which is only possible when `let now = SystemTime::now();` returns a wrong timestamp ‒ implying time went backwards. So from my point of view the error message is correct?

Reply:
Just copy from [here](https://github.com/paritytech/substrate/blob/master/core/consensus/rhd/src/lib.rs#L1305).

Reply:
> As I understand it now.duration_since(UNIX_EPOCH) would only fail if now is smaller than 1970-01-01 00:00:00 UTC which is only possible when let now = SystemTime::now(); returns a wrong timestamp ‒ 

Yeah, that's my intuition as well.

> implying time went backwards. So from my point of view the error message is correct?

Implying that the wall clock on the user machine is set to not current time, which is more plausible to me.

> Just copy from here.

Yeah, this one is better!

(Assuming that we are ok to panic occasionally if the user's environment suddenly returned such date on some OSes that don't use UNIX_EPOCH as the starting point)